[SPARK-38617][SQL][WEBUI] Show Spark rule and phase timings in SQL UI and REST API

yliou opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
The changes proposed are to add a clickable field on SQL UI to show timing of spark phases and rule timing information from QueryPlanningTracker along with a separate REST endpoint to show the timing statistics. Adds a SQLConf to control the number of rules that 
show up on the UI and the REST endpoint call (default value is 15).
UI Change: <img width="310" alt="Screen Shot 2022-05-25 at 11 04 50 AM" src="https://user-images.githubusercontent.com/16739760/170348920-19b13751-a3a1-4df7-8e2a-eb90be246a1d.png">
New REST API endpoint output: 
[SPARK-38617RestOutput.txt](https://github.com/apache/spark/files/8773891/SPARK-38617RestOutput.txt)



<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


### Why are the changes needed?
Currently user is unable to see Spark phase timing information or which Spark rules take the most time for a specific query execution on the UI. This PR adds that information.
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->


### Does this PR introduce _any_ user-facing change?
Yes on the UI
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->


### How was this patch tested?
Tested locally and added a class to test the new endpoint.
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->

yliou committed:
SPARK-38617 show Spark rule and phase timings in SQL UI and REST API

martin-g commented:
[SPARK38617.txt](https://github.com/apache/spark/files/8326833/SPARK38617.txt) is in rich text format. It would be easier to review if it was plain text.

martin-g commented on sql/core/src/test/scala/org/apache/spark/status/api/v1/sql/SqlCompileResourceSuite.scala:
```diff
@@ -0,0 +1,119 @@
+    val salaryDF = salary.withColumnRenamed("personId", "id")
+    val ds = person.join(salaryDF, "id")
+      .groupBy("name", "age", "salary").avg("age", "salary")
+      .filter(_.getAs[Int]("age") <= 30)
```
the filtering filters nothing with the provided two persons. Is this intentional ?

yliou replied:
Yes

ulysses-you commented:
https://github.com/apache/spark/pull/35856 has tried the similar things although at present it miss the details of the compile time for each phase. And its scope also includes the AQE, I think is good to go.

yliou commented on sql/core/src/main/scala/org/apache/spark/status/api/v1/sql/SqlCompilerResource.scala:
```diff
@@ -0,0 +1,82 @@
+    withUI { ui =>
+      val sqlStore = new SQLAppStatusStore(ui.store.store)
+      val appid = ui.store.environmentInfo().sparkProperties.toMap.getOrElse("spark.app.id", "")
+      sqlStore.executionsList(offset, length).map { exec =>
```
It shouldn't be needed. I added SqlCompilerResource.scala based on SqlResource.scala

yliou commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusStore.scala:
```diff
@@ -79,6 +80,14 @@ class SQLAppStatusStore(
+    try {
+      store.read(classOf[CompilerStats], executionId).queryStats
+    } catch {
+      case _: NoSuchElementException => new QueryPlanningTracker()
```
I added this try catch statement as a precaution after adding the if statement on SQLAppStatusListener.scala:393. Some unit tests were failing without the if statement. I can remove the try catch statement if desired.

yliou commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala:
```diff
@@ -160,4 +162,23 @@ class QueryPlanningTracker {
     }
   }
 
+  override def toString(): String = {
```
Renamed method to compileStatsString

yliou commented:
> [SPARK38617.txt](https://github.com/apache/spark/files/8326833/SPARK38617.txt) is in rich text format. It would be easier to review if it was plain text.

Attached new file with plain text format.

yliou commented:
> #35856 has tried the similar things although at present it miss the details of the compile time for each phase. And its scope also includes the AQE, I think is good to go.

What does good to go mean here as I don't see an approval at #35856? From what I see the changes at #35856 are different and complimentary with this PR as that PR adds information regarding AQE plan description changes and this PR has timing of spark phases and rule timing information.

sigmod commented:
cc @jchen5

ulysses-you commented:
@yliou  I agree both can help for troubleshooting. And I have mentioned the difference, `it miss the details of the compile time for each phase`. That PR introduces a disk-based store so I'm thinking if can re-use the same code path. IMO it would be better to go with that PR first at least.

mridulm commented:
+CC @shardulm94, @thejdeep

sigmod commented:
Can we use a more meaning JSON field names? E.g.,
`value` -> `timeMs`
`name` -> `phase`

 ```
{
    “name” : “optimization”,
    “value” : “373"
  }, {
    “name” : “analysis”,
    “value” : “12"
  } ],
  “ruleInfo” : [ {
    “ruleName” : “org.apache.spark.sql.catalyst.optimizer.ConvertToLocalRelation”,
    “value” : “32.659566”,
    “numInvocations” : 5,
    “numEffectiveInvocations” : 3
  }, 
  ....
]

```

yliou commented:
@ulysses-you got it. I think it is possible to re-use the same code path but not sure how that'd work with ExecutionPage.scala to get the info on the UI.

yliou committed:
SPARK-38617 rename added REST fields

yliou commented:
> Can we use a more meaningful JSON field names? E.g., `value` -> `timeMs` `name` -> `phase`
> 
> ```
> {
>    “phase” : “optimization”,
>    “timeMs” : “373"
>  }, {
>    “phase” : “analysis”,
>    “timeMs” : “12"
>  } ],
>  “ruleInfo” : [ {
>    “ruleName” : “org.apache.spark.sql.catalyst.optimizer.ConvertToLocalRelation”,
>    “timeMs” : “32.659566”,
>    “numInvocations” : 5,
>    “numEffectiveInvocations” : 3
>  }, 
>  ....
> ]
> ```
> 
> Also, can `timeMs` be a number instead of a string?

Updated Json field names and changed `timeMs` to be a number

linhongliu-db commented:
@ulysses-you @yliou adding debug information is always good. But I'm worried about two issues:
1. we should be careful when adding information to live UI's in-memory KV store. sooner or later, it will bring headaches
2. we should be careful when adding new REST API endpoints. since two PRs both for troubleshooting, maybe we can use the same endpoint.

yliou committed:
SPARK-38617 rename SqlCompilerResourceSuite and save String names for REST/UI Output

yliou commented on sql/core/src/test/scala/org/apache/spark/status/api/v1/sql/SqlCompileResourceSuite.scala:
```diff
@@ -0,0 +1,119 @@
+/**
+ * Sql Resource Public API Unit Tests running query and extracting the metrics.
+ */
+class SqlCompileResourceSuite
```
Thanks for catching the typo! I renamed the file.

yliou committed:
SPARK-38617 lower SQLConf default and minor cleanup

yliou commented:
Updated the PR to be more concise regarding the names of rules so that Spark event logs aren't as impacted as before in terms off size.

yliou committed:
SPARK-38617 Minor refactoring and rewording of API class names

yliou commented:
@dependabot reopen

yliou commented:
Hi @ulysses-you @linhongliu-db now that https://github.com/apache/spark/pull/35856 got reverted and https://github.com/apache/spark/pull/38567 got merged, should I create another pull request for this feature to try to get it merged?
